src/core/utils/Ledger.ts

```diff
@@ -0,0 +1,230 @@
+        const signingBytes = networkGenerationHash + rawPayload.slice(216);
+        const rawTx = Buffer.from(signingBytes, 'hex');
+        const response = await this.ledgerMessageHandler(path, rawTx, false);
+        // Response from Ledger
+        const h = response.toString('hex');
+        const signature = h.slice(0, 128);
+        const payload = rawPayload.slice(0, 16) + signature + signerPublicKey + rawPayload.slice(16 + 128 + 64, rawPayload.length);
+        const generationHashBytes = Array.from(Convert.hexToUint8(networkGenerationHash));
+        const transactionHash = Transaction.createTransactionHash(payload, generationHashBytes);
+        const signedTransaction = new SignedTransaction(
+            payload,
+            transactionHash,
+            signerPublicKey,
+            transaction.type,
+            transaction.networkType,
+        );
+        return signedTransaction;
+    }
+
+    /**
+     * sign a Symbol Cosignature transaction with a given BIP 44 path
+     *
+     * @param path a path in BIP 44 format
+     * @param cosignatureTransaction a cosinature transaction needs to be signed
+     * @param signerPublicKey the public key of signer
+     * @return a Signed Cosignature Transaction which is signed by account at path on Ledger
+     */
+    public async signCosignatureTransaction(path: string, cosignatureTransaction: AggregateTransaction, signerPublicKey: string) {
+        const rawPayload = cosignatureTransaction.serialize();
+        const signingBytes = cosignatureTransaction.transactionInfo.hash + rawPayload.slice(216);
+        const rawTx = Buffer.from(signingBytes, 'hex');
+        const response = await this.ledgerMessageHandler(path, rawTx, false);
+        // Response from Ledger
+        const h = response.toString('hex');
+        const signature = h.slice(0, 128);
+        const cosignatureSignedTransaction = new CosignatureSignedTransaction(
+            cosignatureTransaction.transactionInfo.hash,
+            signature,
+            signerPublicKey,
+        );
+        return cosignatureSignedTransaction;
+    }
+    /**
+     * handle sending and receiving packages between Ledger and Wallet
+     * @param path a path in BIP 44 format
+     * @param rawTx a raw payload transaction hex string
+     * @param chainCode optionally enable or not the chainCode request
+     * @returns respond package from Ledger
+     */
+    private async ledgerMessageHandler(path: string, rawTx: Buffer, chainCode: boolean) {
+        const TX_INS_FIELD = 0x04;
+        const MAX_CHUNK_SIZE = 255;
+        const CONTINUE_SENDING = '0x9000';
+
+        const curveMask = 0x80; // use Curve25519
+        const bipPath = BIPPath.fromString(path).toPathArray();
+        const apduArray = [];
+        let offset = 0;
+
+        while (offset !== rawTx.length) {
+            const maxChunkSize = offset === 0 ? MAX_CHUNK_SIZE - 1 - bipPath.length * 4 : MAX_CHUNK_SIZE;
```

Comment:
@TriPhan96 `offset` always 0? meaning no need for ternary operator?

Reply:
Hi, thanks @bassemmagdy.  Since the _offset_ variable value will be changed in the while loop, so the conditional operators with it make sense